renderer/backend/InteractionStage.cpp

```diff
@@ -0,0 +1,560 @@
+	if (GLAD_GL_ARB_bindless_texture) {
+		bindlessAmbientInteractionShader = programManager->LoadFromGenerator( "interaction_ambient_bindless", 
+			[this](GLSLProgram *shader) { LoadInteractionShader( shader, "interaction.ambient", true ); } );
+		bindlessStencilInteractionShader = programManager->LoadFromGenerator( "interaction_stencil_bindless", 
+			[this](GLSLProgram *shader) { LoadInteractionShader( shader, "interaction.stencil", true ); } );
+		bindlessShadowMapInteractionShader = programManager->LoadFromGenerator( "interaction_shadowmap_bindless", 
+			[this](GLSLProgram *shader) { LoadInteractionShader( shader, "interaction.shadowmap", true ); } );
+	}
+
+	qglGenBuffers( 1, &poissonSamplesUbo );
+	qglBindBuffer( GL_UNIFORM_BUFFER, poissonSamplesUbo );
+	qglBufferData( GL_UNIFORM_BUFFER, 150 * sizeof(idVec2), nullptr, GL_STATIC_DRAW );
+}
+
+void InteractionStage::Shutdown() {
+	qglDeleteBuffers( 1, &poissonSamplesUbo );
+	poissonSamplesUbo = 0;
+	poissonSamples.Clear();
+}
+
+void InteractionStage::DrawInteractions( viewLight_t *vLight, const drawSurf_t *interactionSurfs ) {
+	if ( !interactionSurfs ) {
+		return;
+	}
+	if ( vLight->lightShader->IsAmbientLight() ) {
+		if ( r_skipAmbient.GetInteger() & 2 )
+			return;
+	} else if ( r_skipInteractions.GetBool() ) {
+		return;
+	}
+
+	GL_PROFILE( "DrawInteractions" );
+
+	PreparePoissonSamples();
+
+	// if using float buffers, alpha values are not clamped and can stack up quite high, since most interactions add 1 to its value
+	// this in turn causes issues with some shader stage materials that use DST_ALPHA blending.
+	// masking the alpha channel for interactions seems to fix those issues, but only do it for float buffers in case it has
+	// unwanted side effects
+	int alphaMask = r_fboColorBits.GetInteger() == 64 ? GLS_ALPHAMASK : 0;
+
+	// perform setup here that will be constant for all interactions
+	GL_State( GLS_SRCBLEND_ONE | GLS_DSTBLEND_ONE | GLS_DEPTHMASK | alphaMask | backEnd.depthFunc );
+	backEnd.currentSpace = NULL; // ambient/interaction shaders conflict
+
+	if ( r_useScissor.GetBool() && !backEnd.currentScissor.Equals( vLight->scissorRect ) ) {
+		backEnd.currentScissor = vLight->scissorRect;
+		GL_ScissorVidSize( backEnd.viewDef->viewport.x1 + backEnd.currentScissor.x1,
+		            backEnd.viewDef->viewport.y1 + backEnd.currentScissor.y1,
+		            backEnd.currentScissor.x2 + 1 - backEnd.currentScissor.x1,
+		            backEnd.currentScissor.y2 + 1 - backEnd.currentScissor.y1 );
+		GL_CheckErrors();
+	}
+
+	// bind the vertex and fragment program
+	ChooseInteractionProgram( vLight, interactionSurfs == backEnd.vLight->translucentInteractions );
+	InteractionUniforms *uniforms = interactionShader->GetUniformGroup<InteractionUniforms>();
+	uniforms->cubic.Set( vLight->lightShader->IsCubicLight() ? 1 : 0 );
+	uniforms->globalLightOrigin.Set( vLight->globalLightOrigin );
+
+	std::vector<const drawSurf_t*> drawSurfs;
```

Comment:
Should it be a member to avoid allocations?

Reply:
Does not make a difference that I could measure, so I'd leave it here. But I've changed it to idList.